plugins/flow-visibility/policy-recommendation/policy_recommendation_job.py

```diff
@@ -0,0 +1,577 @@
+import antrea_crd
+import datetime
+import json
+import kubernetes.client
+import random
+import string
+import uuid
+import yaml
+from ipaddress import ip_address, IPv4Address
+from pyspark.sql import SparkSession
+from pyspark.sql.functions import udf
+from pyspark.sql.types import StringType
+from re import sub
+
+# Column names of flow record table in Clickhouse database used in recommendation job 
+FLOW_TABLE_COLUMNS = [
+    'sourcePodNamespace',
+    'sourcePodLabels',
+    'destinationIP',
+    'destinationPodNamespace',
+    'destinationPodLabels',
+    'destinationServicePortName',
+    'destinationTransportPort',
+    'protocolIdentifier',
+]
+
+NAMESPACE_ALLOW_LIST = [
+    'kube-system',
+    'flow-aggregator',
+    'flow-visibility'
+]
+
+def get_flow_type(destinationServicePortName, destinationPodLabels):
+    if destinationServicePortName != "":
+        return "pod_to_svc"
+    elif destinationPodLabels != "":
+        return "pod_to_pod"
+    else:
+        return "pod_to_external"
+
+def map_flow_to_egress(flow):
+    src = "#".join([flow.sourcePodNamespace, flow.sourcePodLabels])
+    if flow.flowType == "pod_to_external":
+        dst = "#".join([flow.destinationIP, str(flow.destinationTransportPort), flow.protocolIdentifier])
+    else:
+        dst = "#".join([flow.destinationPodNamespace, flow.destinationPodLabels, str(flow.destinationTransportPort), flow.protocolIdentifier])
+    return (src, ("", dst))
+
+def map_flow_to_egress_svc(flow):
+    src = "#".join([flow.sourcePodNamespace, flow.sourcePodLabels])
+    dst = "#".join([flow.destinationServicePortName, str(flow.destinationTransportPort), flow.protocolIdentifier])
+    return (src, dst)
+
+def map_flow_to_ingress(flow):
+    src = "#".join([flow.sourcePodNamespace, flow.sourcePodLabels, str(flow.destinationTransportPort), flow.protocolIdentifier])
+    dst = "#".join([flow.destinationPodNamespace, flow.destinationPodLabels])
+    return (dst, (src, ""))
+
+def combine_network_peers(a, b):
+    if a[0] != "" and b[0] != "":
+        new_src = '|'.join([a[0], b[0]])
+    elif a[0] != "":
+        new_src = a[0]
+    else:
+        new_src = b[0]
+    if a[1] != "" and b[1] != "":
+        new_dst = '|'.join([a[1], b[1]])
+    elif a[1] != "":
+        new_dst = a[1]
+    else:
+        new_dst = b[1]
+    return (new_src, new_dst)
+
+def get_IP_version(IP):
+    return "v4" if type(ip_address(IP)) is IPv4Address else "v6"
+
+def camel(s):
+    s = sub(r"(_|-)+", " ", s).title().replace(" ", "")
+    if not s: return s
+    return s[0].lower() + s[1:]
```

Comment:
I feel many of these utility functions can be moved into a separate file for maintenance.